File: android/app/src/main/kotlin/io/lantern/model/ReplicaModel.kt
package io.lantern.model

import android.app.DownloadManager
import android.content.Context
import android.net.Uri
import android.os.Environment
import android.util.Log
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodCall
import org.getlantern.lantern.MainActivity
import org.getlantern.lantern.R

class ReplicaModel(
    private val activity: MainActivity,
    flutterEngine: FlutterEngine,
) : BaseModel("replica", flutterEngine, masterDB.withSchema("replica")) {

    companion object {
        private const val TAG = "ReplicaModel"
        const val PATH_SEARCH_TERM = "/searchTerm"
        const val PATH_SEARCH_TAB = "/searchTab"
        const val PATH_TO_SHOW_NEW_BADGE = "/showNewBadge"
    }

    init {
        db.mutate { tx ->
            tx.put(PATH_SEARCH_TERM, "")
            tx.put(PATH_SEARCH_TAB, 0)
        }
    }

    override fun doMethodCall(call: MethodCall, notImplemented: () -> Unit): Any? {
        return when (call.method) {
            "downloadFile" -> downloadFile(call)
            "setSuppressUploadWarning" -> setSuppressUploadWarning(call)
            "setSearchTerm" -> {
                db.mutate { tx ->
                    tx.put(PATH_SEARCH_TERM, call.argument<String>("searchTerm")!!)
                }
            }

            "setSearchTab" -> {
                db.mutate { tx ->
                    tx.put(PATH_SEARCH_TAB, call.argument<String>("searchTab")!!)
                }
            }

            "setShowNewBadge" -> {
                db.mutate { tx ->
                    tx.put(PATH_TO_SHOW_NEW_BADGE, call.argument<Boolean>("showNewBadge")!!)
                }
            }

            else -> super.doMethodCall(call, notImplemented)
        }
    }

    private fun downloadFile(call: MethodCall) {
        Log.d(TAG, "downloadFile: ")
        val url = call.argument<String>("url")!!
        val displayName = call.argument<String>("displayName")!!
        val request: DownloadManager.Request = DownloadManager.Request(Uri.parse(url))
        request.setTitle(displayName)
        request.setDescription(activity.getString(R.string.replica_download))
        request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
        // Note - the DownloadManager correctly handles the case where a file with the same name
        // already exists by appending a sequence number to it.
        request.setDestinationInExternalPublicDir(
            Environment.DIRECTORY_DOWNLOADS,
            displayName
        )
        activity.getSystemService(Context.DOWNLOAD_SERVICE)?.let { manager ->
            (manager as DownloadManager).enqueue(request)
        }
    }

    private fun setSuppressUploadWarning(call: MethodCall) {
        val suppress = call.argument<Boolean>("suppress")
        db.mutate { tx ->
            tx.put("suppressUploadWarning", suppress)
        }
    }
}

File: lib/features/replica/models/replica_model.dart
import 'package:lantern/core/utils/common.dart';
import 'package:lantern/features/replica/common.dart';

final replicaModel = ReplicaModel();

class ReplicaModel extends Model {
  ReplicaModel() : super('replica');

  Future<void> downloadFile(
    String url,
    String displayName,
  ) async {
    return methodChannel.invokeMethod('downloadFile', {
      'url': url,
      'displayName': displayName,
    });
  }

  Widget withReplicaApi(ValueWidgetBuilder<ReplicaApi> builder) {
    return sessionModel.replicaAddr((context, replicaAddr, child) {
      return builder(context, ReplicaApi(replicaAddr), child);
    });
  }

  Future<bool?> getSuppressUploadWarning() async {
    return methodChannel.invokeMethod('get', 'suppressUploadWarning');
  }

  Future<void> setSuppressUploadWarning(bool suppress) async {
    return methodChannel.invokeMethod('setSuppressUploadWarning', {
      'suppress': suppress,
    });
  }

  Future<void> setSearchTerm<T>(String searchTerm) async {
    return methodChannel.invokeMethod('setSearchTerm', <String, dynamic>{
      'searchTerm': searchTerm,
    });
  }

  Future<void> setSearchTab<T>(int searchTab) async {
    return methodChannel.invokeMethod('setSearchTab', <String, dynamic>{
      'searchTab': searchTab.toString(),
    });
  }

  Future<void> setShowNewBadge(bool showNewBadge) async {
    return methodChannel.invokeMethod('setShowNewBadge', {
      'showNewBadge': showNewBadge,
    });
  }

  Widget getShowNewBadgeWidget(ValueWidgetBuilder<bool> builder) {
    return subscribedSingleValueBuilder<bool>(
      '/showNewBadge',
      defaultValue: true,
      builder: builder,
    );
  }

  Widget getSearchTermWidget(ValueWidgetBuilder<String> builder) {
    return subscribedSingleValueBuilder<String>(
      '/searchTerm',
      defaultValue: '',
      builder: builder,
    );
  }

  Widget getSearchTabWidget(ValueWidgetBuilder<String> builder) {
    return subscribedSingleValueBuilder<String>(
      '/searchTab',
      defaultValue: '',
      builder: builder,
    );
  }

  Future<bool> getShowNewBadge() async {
    return methodChannel
        .invokeMethod('get', '/showNewBadge')
        .then((value) => value??false);
  }

  Future<String> getSearchTerm() async {
    return methodChannel
        .invokeMethod('get', '/searchTerm')
        .then((value) => value.toString());
  }

  Future<String> getSearchTab() async {
    return methodChannel
        .invokeMethod('get', '/searchTab')
        .then((value) => value.toString());
  }
}
